Remove DRF

kdeloach opened the pull request:
This is an experiment in replacing DRF views with decorated API view functions. The objective is to dumb down the code a little bit and make things explicit. Tests need to be updated.

This PR demonstrates a possible alternative to the DRF approach for writing the API and also demonstrates how to create composable API views.

For example, `/layers.json` `/all/layers.json` and `/user/foo/layers.json` all use the same queryset with different filters applied.

I'd like to get some feedback before continuing research on this.

kdeloach commented on src/rf/apps/home/views.py:
```diff
@@ -3,155 +3,92 @@
-
-    def delete(self, request, *args, **kwargs):
-        return self.destroy(request, *args, **kwargs)
+@csrf_exempt
```
This is for testing only.

kdeloach commented:
@lewfish @lliss Love it / hate it?

lliss commented:
Since I haven't worked on the API that much, I have less skin in the game on this. I do think the patterns you ended up creating in `home/views.py` look pretty nice. The pattern for specifying allowed methods is pretty easy and it is nice how you split the response methods to tiny functions.

Your decorators are nice as well.

I do like how DRF gives you the ability to swap between response types easily on the requester's side. I also have fears that since we're reinventing a bunch of stuff we might miss something that has been pretty well tested and thought out by the DRF team.

If we decide to go this route, I'd be fine with it.

kdeloach commented:
> I do like how DRF gives you the ability to swap between response types easily on the requester's side. 

I was thinking about writing a decorator to do this, but since we use JSON 99% of the time, all requests with the `@HttpApiView` decorator are automatically serialized to JSON *unless* the view returns a custom `HttpResponse` (Ref: https://github.com/azavea/raster-foundry/pull/60/files#diff-6b6b61b55ead82f897c23a5b7c843030R23)

> I also have fears that since we're reinventing a bunch of stuff we might miss something that has been pretty well tested and thought out by the DRF team.

Good point.

kdeloach commented:
Squashed a bunch of changes.

lewfish commented:
I think you have a nice solution, but it doesn't seem all that different than DRF. Some of the decorators you wrote remind me of PermissionClasses and the_get_layer_models function reminds me of get_queryset in DRF. I did find your approach more explicit, but that's only because it's more compact and easy to read -- at the moment. If you kept expanding your solution to handle different situations that come up in different apps we want to build, you'd probably end up with something just as complex as DRF. On the other hand, I think it might be worth using something smaller and more customized like you wrote for a small app, where we don't want to invest too much time in learning a new framework.

kdeloach commented:
Squashed more changes and refactored the backend a little more. You can explore and test the API using the [swagger UI](http://editor.swagger.io/). The spec is here `src/rf/apps/home/static/swagger.yaml`.

kdeloach commented:
@mmcfarland Have a few minutes to review this today?

kdeloach commented:
Rebased onto develop.

mmcfarland commented on src/rf/apps/core/decorators.py:
```diff
@@ -0,0 +1,160 @@
+                                  MethodNotAllowed, Unauthorized)
+from apps.core.models import User
+
+
```
At a glance, it doesn't look like it, but is there much overlap between this and `django-tinsel`?

kdeloach replied:
Besides being decorator-based, I don't think there's much of an overlap. The routing logic that `django-tinsel` performs with decorators exists in the view functions here.

mmcfarland commented:
My comments are pretty similar to those above.  I find it easier to read and more compact than the DRF implementation, just unsure of if that holds up over time.  I suspect for this app, it will, since there isn't much more CRUD to do and additional endpoints will probably more processing/tasking oriented than data manipulation at a REST-level.  If the team if comfortable with it, this approach seems to give a bit more flexibility at the moment.  As @lliss points out, DRF probably gives a lot for free but at the expense of a certain reliance on the external project.  If it's not providing much for our use case, I don't mind dropping it.

kdeloach commented:
Thanks for the feedback.

I think this solution will be more than adequate for this project.

Merging!
